### DISCUSSION TITLE
Can we make `sh` action to capture only `stdout`?

### DISCUSSION THREAD

**Discussion by GRiMe2D** [fastlane/fastlane#17686] [Ideas] [answered]
Currently `sh` command captures both `stderr` and `stdout`.

Can we make it to capture only `stdout` or even event better, return tuple captured `stdout`, `stderr`.

**Comment by ainame** [comment, marked as answer]
This might not be the solution for your case but you can call any commands with the redirection for stderr?

```ruby
stdout = sh('ls 2> /dev/null') # to drop stderr
stderr = sh('ls 1> /dev/null') # or capture only stderr
sh('ls 1> stdout.log 2> stderr.log') # or capture both by log files
```

Additionally, looking at the current implementation, you can then receive the only stdout with `error_callback` block only when getting errors.

```ruby
sh('ls 2> /dev/null', error_callback: -> {|stdout| do_something(stdout) }) 
```

In terms of feasibility of returning tuple, of course, nothing is impossible but I think code will be much more complex to keep maintain existing behaviour.

---------------

This is how capturing works currently. It uses mixed IO object of `stdout/err` We have to use `Open3.popen3`  to separate stdout and stderr.

https://github.com/fastlane/fastlane/blob/2624a24407dee94f67d7e39b5651f9fdc41dcb00/fastlane/lib/fastlane/helper/sh_helper.rb#L54-L61

**Reply by GRiMe2D** [reply]
Thanks for the detailed answer. Currently I solved my problem by piping `stderr` to a temporary file and reading it later.
